### PR TITLE
fix compile error, and some refactors

### PR DESCRIPTION
**Pull request by qinzuoyan**

### CODE REVIEW

**Comment by qinzuoyan** [comment]
Updated:
1. fix bug in DBImpl::ApplyLearningState to use 'auto&' when update new file numbers:
`
  auto& new_files = edit.GetNewFiles();
`
2. when learning sstables, rename 'learn_dir/xxx.sst' to 'data_dir/xxx.sst.learn' to avoid conflicting with exist files.
4. improve setting of LastDurableSequence when do flush job to make the value more correct.
3. add more assert check to validate logic.
5. improve logging.

**Comment by imzhenyu** [comment]
For 2, since we now have two rounds of rename: from learning dir to data dir, and from data dir "_.learn" to data dir "_ (possibly with new file number). It is possible for us to merge these two rounds?
Others look good for me:)

**Comment by qinzuoyan** [comment]
Actually, I have thought the problem when fixing it. To avoid renaming files for two rounds, we can:
1. when copying files from remote, put files directly into data_dir and name it as 'data_dir/xxx.sst.learn', then we need not rename files in rrdb_service_impl::apply_learn_state();
But I didn't do it because I think copying files from remote is long time processing, and put files into separate learn_dir can mostly reduce impact to data_dir.
2. when renaming files in rrdb_service_impl::apply_learn_state(), we can merge assigning new file number and renaming files into one step, then we need not rename files in db_impl::ApplyLearnState().
But I didn't go this way because I think assigning new file number and invoking LogAndApply() should be atomic and protected in one mutex, because apply_learn_state() may be invoked concurrently.

In short, I also think the current solution is not the final, and I have been thinking about the better way.

**Comment by imzhenyu** [comment]
Good - I'm merging it now. Btw, I'm going to push the refactored rDSN with c interface to master today. I'll also make the necessary updates in rocksdb.replicated.

**Comment by qinzuoyan** [comment]
Nice :)
